File: projects/ruby/lib/dao.rb
# frozen_string_literal: true

require 'securerandom'
require_relative 'result_store'
require_relative 'operations'

##
# DAO (Data Access Object) provides a unified interface for operation invocation.
#
# All operations follow an async pattern:
# 1. Operation is invoked with call(operation, params)
# 2. UUID is generated immediately
# 3. Result is stored with status="pending"
# 4. Operation executes (simulated synchronously for demo)
# 5. Result is updated to status="completed" with result data
# 6. Initial response returns {id: uuid, status: "pending"}
#
# Client can poll for results using /result/poll operation.
#
class DAO
  def initialize
    @store = ResultStore.new
  end

  ##
  # Call an operation with parameters.
  #
  # @param operation [String] Hierarchical operation path (e.g., "/prompt/generate")
  # @param params [Hash] Operation-specific parameters
  # @return [Hash] Operation response with :id, :status, and optional :result or :error
  #
  def call(operation, params = {})
    # Special case: /result/poll retrieves existing result
    if operation == '/result/poll'
      poll_id = params['id'] || params[:id]
      raise ArgumentError, 'Missing required parameter: id' if poll_id.nil?

      stored = @store.get(poll_id)
      return { error: "Result not found or expired: #{poll_id}" } if stored.nil?

      return stored
    end

    # Generate correlation UUID
    id = SecureRandom.uuid

    # Route to operation handler
    begin
      result = route_operation(operation, params)

      # Special case: /usage/track completes synchronously
      if operation == '/usage/track'
        completed_response = {
          id: id,
          status: 'completed',
          result: result
        }
        @store.set(id, completed_response)
        return completed_response
      end

      # Standard async pattern: store completed result, return pending
      completed_response = {
        id: id,
        status: 'completed',
        result: result
      }
      @store.set(id, completed_response)

      # Return initial pending response (async pattern)
      {
        id: id,
        status: 'pending'
      }

    rescue StandardError => e
      # Store failed result and return error
      failed_response = {
        id: id,
        status: 'failed',
        error: e.message
      }
      @store.set(id, failed_response)

      { error: e.message }
    end
  end

  private

  ##
  # Route operation to appropriate handler.
  #
  # @param operation [String] Operation path
  # @param params [Hash] Operation parameters
  # @return [Hash] Operation result
  # @raise [ArgumentError] If operation is unknown
  #
  def route_operation(operation, params)
    case operation
    when '/prompt/generate'
      Operations.process_prompt(params)
    when '/template/create'
      Operations.create_template(params)
    when '/template/render'
      Operations.render_template(params)
    when '/usage/track'
      Operations.track_usage(params)
    else
      raise ArgumentError, "Invalid operation: #{operation}"
    end
  end
end

File: projects/javascript/lib/dao.js
/**
 * DAO - Data Access Object providing unified interface
 *
 * All operations follow an async pattern:
 * 1. Generate UUID immediately
 * 2. Store result with status="pending"
 * 3. Execute operation (simulated synchronously for demo)
 * 4. Update result to status="completed"
 * 5. Return initial response {id: uuid, status: "pending"}
 *
 * Clients poll for results using /result/poll operation.
 */

import { v4 as uuidv4 } from 'uuid';
import ResultStore from './store.js';
import * as operations from './operations.js';

class DAO {
  constructor() {
    this.store = new ResultStore();
  }

  /**
   * Call an operation with parameters
   * @param {string} operation - Hierarchical operation path
   * @param {object} params - Operation-specific parameters
   * @returns {object} Operation response with id, status, and optional result or error
   */
  call(operation, params = {}) {
    // Special case: /result/poll retrieves existing result
    if (operation === '/result/poll') {
      const pollId = params.id;
      if (!pollId) {
        return { error: 'Missing required parameter: id' };
      }

      const stored = this.store.get(pollId);
      if (!stored) {
        return { error: `Result not found or expired: ${pollId}` };
      }

      return stored;
    }

    // Generate correlation UUID
    const id = uuidv4();

    // Route to operation handler
    try {
      const result = this._routeOperation(operation, params);

      // Special case: /usage/track completes synchronously
      if (operation === '/usage/track') {
        const completedResponse = {
          id: id,
          status: 'completed',
          result: result
        };
        this.store.set(id, completedResponse);
        return completedResponse;
      }

      // Standard async pattern: store completed result, return pending
      const completedResponse = {
        id: id,
        status: 'completed',
        result: result
      };
      this.store.set(id, completedResponse);

      // Return initial pending response (async pattern)
      return {
        id: id,
        status: 'pending'
      };

    } catch (error) {
      // Store failed result and return error
      const failedResponse = {
        id: id,
        status: 'failed',
        error: error.message
      };
      this.store.set(id, failedResponse);

      return { error: error.message };
    }
  }

  /**
   * Route operation to appropriate handler
   * @param {string} operation - Operation path
   * @param {object} params - Operation parameters
   * @returns {object} Operation result
   * @private
   */
  _routeOperation(operation, params) {
    switch (operation) {
      case '/prompt/generate':
        return operations.processPrompt(params);
      case '/template/create':
        return operations.createTemplate(params);
      case '/template/render':
        return operations.renderTemplate(params);
      case '/usage/track':
        return operations.trackUsage(params);
      default:
        throw new Error(`Invalid operation: ${operation}`);
    }
  }
}

export default DAO;
